MuonAnalysis/MuonAssociators/plugins/MuonMCClassifier.cc

```diff
@@ -291,87 +301,166 @@ MuonMCClassifier::produce(edm::Event& iEvent, const edm::EventSetup& iSetup)
                     edm::LogWarning("MuonMCClassifier") << "\n***WARNING:  This I do NOT understand: why no match back in UpdSTA? *** \n";
                 }
             }
-        }
+        } else {
+            edm::LogVerbatim("MuonMCClassifier") <<"\t No matching TrackingParticle is found ";
+	}
+
         if (tp.isNonnull()) {
             bool isGhost = muMatchBack != mu;
-            if (isGhost) edm::LogVerbatim("MuonMCClassifier") <<"\t This seems a GHOST ! classif[i] will be < 0";
+            if (isGhost) edm::LogVerbatim("MuonMCClassifier") <<"\t *** This seems a Duplicate muon ! classif[i] will be < 0 ***";
+
+	    // identify signal and pileup TP
+	    tpBx[i]      = tp->eventId().bunchCrossing();
+	    tpEv[i]      = tp->eventId().event();
 
             hitsPdgId[i] = tp->pdgId();
             prodRho[i]   = tp->vertex().Rho();
             prodZ[i]     = tp->vertex().Z();
-	    edm::LogVerbatim("MuonMCClassifier") <<"\t TP pdgId = "<<hitsPdgId[i] << ", vertex rho = " << prodRho[i] << ", z = " << prodZ[i];
+
+	    // added info on GEANT process producing the TrackingParticle
+	    std::vector<SimVertex> G4Vs = tp->parentVertex()->g4Vertices();
```

Comment:
hi @abbiendi - this can be const std::vector<SimVertex> &G4Vs to avoid an unneeded copy